larcorealg/Geometry/LineClosestPoint.tcc

```diff
@@ -0,0 +1,129 @@
+/**
+ * @file   larcorealg/Geometry/LineClosestPoint.tcc
+ * @brief  Utility for intersection of two 3D lines: template implementation.
+ * @author Gianluca Petrillo (petrillo@slac.stanford.edu)
+ * @see    larcorealg/Geometry/LineClosestPoint.h
+ */
+
+#ifndef LARCOREALG_GEOMETRY_LINECLOSESTPOINT_H
+# error "LineClosestPoint.tcc should not be included directly: #include \"larcorealg/Geometry/LineClosestPoint.h\" instead."
+#endif
+
+
+// LArSoft and framework libraries
+#include "cetlib/pow.h" // cet::square
+
+// C/C++ libraries
+#include "larcorealg/Geometry/geo_vectors_utils.h" // geo::vect::dot()
+
+// C++ standard library
+#include <cmath> // std::abs()
+#include <cassert>
+
+
+// -----------------------------------------------------------------------------
+template <typename Point, typename Vector>
+Point geo::LineClosestPoint(
+  Point const& startA, Vector const& dirA,
+  Point const& startB, Vector const& dirB,
+  std::pair<double, double>* locOnLines /* = nullptr */
+) {
+  /*
+   * The point on the first line ("A"):
+   *
+   *     p1(t) = c1 + t w1 (c1 the starting point, w1 its direction)
+   *
+   * has the minimal distance from the other line (c2 + u w2, with the same
+   * notation) at
+   *
+   *     t = [-(dc,w1)(w2,w2) + (dc,w2)(w1,w2)] / [ (w1,w2)^2 - (w1,w1)(w2,w2) ]
+   *     u = [ (dc,w2)(w2,w2) - (dc,w1)(w1,w2)] / [ (w1,w2)^2 - (w1,w1)(w2,w2) ]
+   *
+   * (where (a,b) is a scalar product and dc = (c1 - c2) ).
+   * If w1 and w2 are unit vectors, t and u are in fact the distance of the
+   * point from the "start" of the respective lines in "standard" geometry
+   * units.
+   * 
+   * If `locOnLines` is non-null, it is filled with { t, u }.
+   */
+
+  // aliases for quick notation
+  Point const& c1 = startA;
+  Vector const& w1 = dirA;
```

Comment:
I am not sure readability benefits from it, but ok.

BTW: you can use the "suggestion" feature of GitHub for this type of comments (I think it's enough to put `suggesion` after the opening triple back tick):
```suggestion
  auto const& [c1, w1] = std::tie(startA, dirA);
```
Why? because it looks cool.

Reply:
Yes, it's cool. 😃